### ISSUE TITLE
tablet primary replica selection may be biased

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylladb#17752]
Currently, when a tablets-enabled table is created, the replicas for each tablet are determined by the network-topology strategy natural_endpoints_tracker, using the token metadata ring order.
The first replica for each tablet is later used by repair_tablets with the primary-ranges option and when no specific ranges are given.  Therefore, we would like the primary replica for each tablet to be distributed uniformly in the cluster so to balance the load.
The initial selection, using the natural_endpoints_tracker should be fine with respect to selecting the primary replica, however, tablet splitting maintains the same order of replicas for the split tablets, so we may end up with a bias if we start with a small number of initial tablets  (a single tablet in the pathological case).

@avikivity proposed to use a modulo function of the tablet id to rotate the order of the tablet replicas to balance the load on primary replicas.  To do that, we'll need to get tablet_info.replicas using a function, taking the tablet_id as a parameter.

Also, its tablet_info::operator== today compares the replicas as an ordered vector.
@tgrabiec is that on purpose or should it be implemented using e.g. an unordered_set so not be sensitive to replicas order?

@nyh / @tgrabiec / @raphaelsc: with respect to base/view tablet replica pairing, how do we deal with tablet split and merge?
Do we trigger a split for views when we decide to split the base table?
What's the plan for merge?
Also, even if we trigger split/merge at the same time, I don't think we ensure that the base table tablets are split in tandem with view replicas as tablet splitting is autonomous for each tablet.  Would rotating the replica order after a split hamper with base/view tablet pairing?
Ref https://github.com/scylladb/scylladb/issues/17147

**Comment by bhalevy** [comment]
Alternatively, we can start with rotating the replicas order only in `repair_service::repair_tablets` to balance the load.
That would still be sensitive to load balancing and split/merge if those are not paused by the user prior to invoking repair on each node, but that should be solved with native tablets repair that will schedule repair for each tablet replica set, picking the repair-master in the process arbitrarily, not necessarily related to the primary replica is it's defined today.

**Comment by tgrabiec** [comment]
>Also, its tablet_info::operator== today compares the replicas as an ordered vector.
>@tgrabiec is that on purpose or should it be implemented using e.g. an unordered_set so not be sensitive to replicas order?

It's used in persistence tests currently, so strict comparison is preferred. 

> @nyh / @tgrabiec / @raphaelsc: with respect to base/view tablet replica pairing, how do we deal with tablet split and merge?
Do we trigger a split for views when we decide to split the base table?
What's the plan for merge?

We don't maintain same tablet mapping (nor the count) between base and view tables so they can be split and merged independently. Tablets are migrated independently. Logical pairing is maintained by migration, but physically the base replica and view replica can live on different nodes. 

> Would rotating the replica order after a split hamper with base/view tablet pairing?

Yes.

**Comment by avikivity** [comment]
We shouldn't rotate the replica order. Instead, primary owner = natural_endpoints[token][nr_tablets_before_this_token % full_replication_factor]

So, if we have three racks a, b, c then the first tablet's owner is on rack a, the second on rack b, the third on rack c, the fourth back on rack a. When we split or merge, the owners get reshuffled, but they also get (a little) reshuffled when a node is bootstrapped or decommissioned with vnodes.

**Comment by nyh** [comment]
> @nyh / @tgrabiec / @raphaelsc: with respect to base/view tablet replica pairing, how do we deal with tablet split and merge? Do we trigger a split for views when we decide to split the base table?

No. We may in some cases (see https://github.com/scylladb/scylladb/issues/17043) but not in the general case. Consider that in the general case, the size of the base table and view are not the same, and so are their popularity, so we may not even want them to always have the same number of tablets.

The important thing is that if we split a view tablet (for example), and a specific replica of this tablet was second (say) in this tablet's list of replicas, then after the split each of the halves of this replica must remain the second in their new lists.

> What's the plan for merge? Also, even if we trigger split/merge at the same time, I don't think we ensure that the base table tablets are split in tandem with view replicas as tablet splitting is autonomous for each tablet. Would rotating the replica order after a split hamper with base/view tablet pairing? Ref #17147

Merge is indeed a complication - see https://github.com/scylladb/scylladb/issues/17265.

**Comment by nyh** [comment]
> We shouldn't rotate the replica order. Instead, primary owner = natural_endpoints[token][nr_tablets_before_this_token % full_replication_factor]
> 
> So, if we have three racks a, b, c then the first tablet's owner is on rack a, the second on rack b, the third on rack c, the fourth back on rack a. When we split or merge, the owners get reshuffled, but they also get (a little) reshuffled when a node is bootstrapped or decommissioned with vnodes.

If the primary replica get reshuffled, it's not great for repair or Alternator TTL, which use the primary replica to avoid doing the same work 3 times. E.g., consider "repair -pr", you run it on node A and it repairs a certain range because it is a primary range there. Then, your run the primary-range repair or node B, but between those times some split happened and caused us to suddenly decide that that range we already repaired on A, is not B's primary range and it will be repaired for a second time on B.

I have an idea: We could we make the choice of primary replica a hash function of the the tablet's **definition** - e.g., the primary replica in a DC will be replica number `hash(token_start, token_end) % RF`.  This hash function will not change if **other** tablets get split or merged. It will change if the tablet itself gets split or merged, but I think this is fine - again, the only purpose of this "primary owner" will be to load-balance work that needs to happen only on one replica and not on all RF - like repair or Alternator TTL.

**Comment by asias** [comment]
> Alternatively, we can start with rotating the replicas order only in `repair_service::repair_tablets` to balance the load. That would still be sensitive to load balancing and split/merge if those are not paused by the user prior to invoking repair on each node, but that should be solved with native tablets repair that will schedule repair for each tablet replica set, picking the repair-master in the process arbitrarily, not necessarily related to the primary replica is it's defined today.
> 
> Cc @asias

Selecting which node to act as the repair master for a given tablet (token range) using primary replica is only for compatibility.  When scheduling repair cluster wide, we have much better ways than using the primary replica. Even if some nodes have more primary replicas, it does not  mean we would overload them in terms of repair, because repair will work on all replicas anyway, no matter who is the repair master. Yes, repair master would have a bit more work but that is not like only the repair master would do the work. 

So I do not think it is worth complicating the primary replica with rotation.
